core/src/apps/cardano/sign_tx.py

```diff
@@ -67,522 +90,563 @@
+        total_amount_to_show += amount_to_show or 0
 
+        await _process_asset_groups(ctx, msg, builder, output.asset_groups_count)
 
-def _cborize_tx_body(keychain: seed.Keychain, msg: CardanoSignTx) -> dict:
-    inputs_for_cbor = _cborize_inputs(msg.inputs)
-    outputs_for_cbor = _cborize_outputs(
-        keychain, msg.outputs, msg.protocol_magic, msg.network_id
-    )
+    builder.finish_outputs()
 
-    tx_body = {
-        0: inputs_for_cbor,
-        1: outputs_for_cbor,
-        2: msg.fee,
-    }
+    if total_amount > LOVELACE_MAX_SUPPLY:
+        raise wire.ProcessError("Total transaction amount is out of range!")
 
-    if msg.ttl:
-        tx_body[3] = msg.ttl
+    return total_amount_to_show
 
-    if msg.certificates:
-        certificates_for_cbor = _cborize_certificates(keychain, msg.certificates)
-        tx_body[4] = certificates_for_cbor
 
-    if msg.withdrawals:
-        withdrawals_for_cbor = _cborize_withdrawals(
-            keychain, msg.withdrawals, msg.protocol_magic, msg.network_id
+async def _process_asset_groups(
+    ctx: wire.Context,
+    msg: CardanoSignTxInit,
+    builder: TxBuilder,
+    asset_groups_count: int,
+) -> None:
+    """Read, validate and serialize the asset groups of an output."""
+    if asset_groups_count == 0:
+        return
+    previous_policy_id: bytes | None = None
+    for _ in range(asset_groups_count):
+        asset_group: CardanoAssetGroup = await ctx.call(
+            CardanoTxItemAck(), CardanoAssetGroup
         )
-        tx_body[5] = withdrawals_for_cbor
+        asset_group.policy_id = bytes(asset_group.policy_id)
+        _validate_asset_group(previous_policy_id, asset_group)
+        builder.add_asset_group(asset_group.policy_id, asset_group.tokens_count)
 
-    # tx_body[6] is for protocol updates, which we don't support
+        previous_policy_id = asset_group.policy_id
 
-    if msg.auxiliary_data:
-        auxiliary_data_cbor = get_auxiliary_data_cbor(
-            keychain, msg.auxiliary_data, msg.protocol_magic, msg.network_id
+        await _process_tokens(
+            ctx, msg, builder, previous_policy_id, asset_group.tokens_count
         )
-        tx_body[7] = hash_auxiliary_data(bytes(auxiliary_data_cbor))
-
-    if msg.validity_interval_start:
-        tx_body[8] = msg.validity_interval_start
+    builder.finish_asset_groups()
 
-    return tx_body
 
+async def _process_tokens(
+    ctx: wire.Context,
+    msg: CardanoSignTxInit,
+    builder: TxBuilder,
+    policy_id: bytes,
+    tokens_count: int,
+) -> None:
+    """Read, validate, confirm and serialize the tokens of an asset group."""
+    previous_asset_name_bytes: bytes | None = None
+    for _ in range(tokens_count):
+        token: CardanoToken = await ctx.call(CardanoTxItemAck(), CardanoToken)
+        token.asset_name_bytes = bytes(token.asset_name_bytes)
+        _validate_token(previous_asset_name_bytes, token)
+        if msg.signing_mode == CardanoTxSigningMode.ORDINARY_TRANSACTION:
```

Comment:
I just saw we only call `_show_outputs` from `_show_standard_tx`, not from `_show_stake_pool_registration_tx`. Even though pool registrations do contain outputs.
https://github.com/vacuumlabs/trezor-firmware/blob/40bd66ded4132121c0ffb10fd6be7d47a050b323/core/src/apps/cardano/sign_tx.py#L556

But I see I've lost this condition from `_process_outputs` anyway. So what's the correct behavior? Show all the output-related screens regardless of signing mode? Was this a bug in the old version then?

Reply:
Hm I see. So I think that pool registration as owner is only about signing as owner. So you are not really worried about the outputs because you are not spending anything - the pool operator is the owner of the inputs, not the pool owners. So I guess the `ifs` are supposed to be there both in `_process_outputs` and in `_process_tokens` so we don't show outputs and tokens in case of a pool registration. @refi93 should confirm this please :wink: